jdaviz/app.py

```diff
@@ -1042,23 +1024,34 @@ def get_sub_regions(self, subset_state, simplify_spectral=True):
                         return new_spec
                     else:
                         if isinstance(two, list):
-                            # two[0]['glue_state'] = subset_state.state2.__class__.__name__
                             two[0]['glue_state'] = "AndNotState"
                         # Return two first so that we preserve the chronology of how
                         # subset regions are applied.
                         return one + two
                 elif subset_state.op is operator.and_:
                     # This covers the AND subset mode
 
-                    # Example of how this works:
-                    # a = SpectralRegion(4 * u.um, 7 * u.um)
-                    # b = SpectralRegion(5 * u.um, 6 * u.um)
+                    # Example of how this works with "one" being the AND region
+                    # and "two" being two Range subsets connected by an OR state:
+                    # one = SpectralRegion(4.5 * u.um, 7.5 * u.um)
+                    # two = SpectralRegion(4 * u.um, 5 * u.um) + SpectralRegion(7 * u.um, 8 * u.um)
                     #
-                    # b.invert(a.lower, a.upper)
+                    # oppo = two.invert(one.lower, one.upper)
+                    # Spectral Region, 1 sub-regions:
+                    #   (5.0 um, 7.0 um)
+                    #
+                    # oppo.invert(one.lower, one.upper)
                     # Spectral Region, 2 sub-regions:
-                    #   (4.0 um, 5.0 um)   (6.0 um, 7.0 um)
+                    #   (4.5 um, 5.0 um)   (7.0 um, 7.5 um)
                     if isinstance(two, SpectralRegion):
-                        return two.invert(one.lower, one.upper)
+                        # Taking an AND state of an empty region is allowed
+                        # but there is no way for SpectralRegion to display that information.
+                        # Instead, we raise a ValueError
+                        if one.upper.value < two.lower.value or one.lower.value > two.upper.value:
+                            raise ValueError("AND mode should overlap with existing subset")
```

Comment:
how would this get triggered?  I tried creating this situation in specviz and it displays the state as I'd expect in the plugin and just doesn't highlight anything in the viewer (as expected), but I didn't see any traceback.  Is this not needed?  Or is the traceback caught somewhere?

Reply:
Hmm, thought I wrote a test for this one. Let me look at this for a bit and get back to you.

Reply:
I checked, the problem is: what should be returned from the result of applying Range, then OR, then AND on an empty region (i.e. not where the two subregions are)? Ideally it should be empty but we cannot send an empty `SpectralRegion` back up the recursive tree. I chose to do this instead since why would someone AND an empty region?

I agree this is an issue but I'm not sure at the moment how to resolve it.

Reply:
Is there something equivalent to "identity" where it is not empty but it also won't change the result? But I agree it might be out of scope to fix here.

Reply:
When you say "identity", what do you mean? I couldn't find anything that achieved that goal but I agree that should be out of scope for this PR.